Add Unify Circuit

braam opened the pull request:
…nents files can be found in homeassistant/components/circuit. Also changed the .coveragerc file and CODEOWNERS is updated. de requirements_all.txt also updated by running the script. The platform is only configurable through configuration.yaml, everything was tested live and works fine. Documentation will follow once it's accepted.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
circuit:
  webhook:
    - name: circuit_hass
      url: https://eu.yourcircuit.com/rest/v2/webhooks/incoming/<ID>
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/13584

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

braam committed:
Added new notification platform for Unify Circuit webhooks. New components files can be found in homeassistant/components/circuit. Also changed the .coveragerc file and CODEOWNERS is updated. de requirements_all.txt also updated by running the script. The platform is only configurable through configuration.yaml, everything was tested live and works fine. Documentation will follow once it's accepted.

homeassistant commented:
Hi @braam, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#35756).

Once you do that we will be able to review and accept this pull request.

Thanks!

MartinHjelmare commented on homeassistant/components/circuit/notify.py:
```diff
@@ -0,0 +1,43 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({vol.Required(CONF_URL): cv.url})
```
Please set up the integration from a config schema in the integration init module. Use a `CONFIG_SCHEMA`, and load the notify platform via the discovery helper `load_platform`.

See other integrations that uses `load_platform` for examples.

Also see ADR 7:
https://github.com/home-assistant/architecture/blob/master/adr/0007-integration-config-yaml-structure.md

braam replied:
Thank you for your suggestion but I think it's not necessary as the component only send outgoing messages and do not accept incoming messages. Circuit from Unify is actually almost the same as MS Teams. So my code is heavily based on the msteams component.

For more than one webhook I tested successfully:
notify:
  - platform: circuit
    name: circuit_webhook_1
    url: https://xxx/ID_1
  - platform: circuit
    name: circuit_webhook_2
    url: https://xxx/ID_2

MartinHjelmare replied:
We won't accept this PR in its current state. Please follow my request above.

braam replied:
Thanks for your quick reply, I looked into the documentation and found this example: https://github.com/home-assistant/example-custom-config/tree/master/custom_components/example_load_platform/ 

Is that what you are requesting? The configuration.yaml will then be changed to:
circuit:
  name: circuit_webhook_1
  url: https://xxx/ID

MartinHjelmare replied:
Yes. Make sure to add a `CONFIG_SCHEMA` in the init module of the integration. That's not in the example.

braam replied:
New commit added with code changes as requested. I was sruggling a bit with the logic to handle this in the configuration.yaml file. I ended up with new config layout:

```
# Example configuration.yaml entry
circuit:
  - webhook:
      - name: webhook_1
        url: "https://eu.yourcircuit.com/rest/v2/webhooks/incoming/<ID-1>"
      - name: webhook_2
        url: "https://eu.yourcircuit.com/rest/v2/webhooks/incoming/<ID-2>"
      - name: webhook_multiple
        url: 
          - "https://eu.yourcircuit.com/rest/v2/webhooks/incoming/<ID-1>"
          - "https://eu.yourcircuit.com/rest/v2/webhooks/incoming/<ID-2>"   
```

braam committed:
Added CONFIG_SCHEMA and _WEBHOOK_SCHEMA and load_platform inside the __init__ file. Also changed the notify.py file for catching up with the new config style. Config checks with black, flake8, isort and lint tests, all get PASS.

MartinHjelmare commented on homeassistant/components/circuit/__init__.py:
```diff
@@ -0,0 +1,53 @@
+    {
+        DOMAIN: vol.All(
+            cv.ensure_list,
+            [
```
Do we need a list here and a list for the webhook schema?

braam replied:
The list for the CONF_URL is needed so we can submit the same notify message to multiple conversations within Circuit. The list for the CONFIG_SCHEMA I'll change that. New commit in PR will be submitted.

See configuration.yaml entries to clarify
```
# Example configuration.yaml entry
circuit:
  - webhook:
      - name: webhook_1
        url: https://eu.yourcircuit.com/rest/v2/webhooks/incoming/<ID-1>
      - name: webhook_2
        url: https://eu.yourcircuit.com/rest/v2/webhooks/incoming/<ID-2>
      - name: webhook_multiple
        url: 
          - https://eu.yourcircuit.com/rest/v2/webhooks/incoming/<ID-1>
          - https://eu.yourcircuit.com/rest/v2/webhooks/incoming/<ID-2>
```

MartinHjelmare commented on homeassistant/components/circuit/__init__.py:
```diff
@@ -0,0 +1,53 @@
+    domain_config = config.get(DOMAIN, [])
+
+    for conf in domain_config:
+        for webhook_conf in conf.get(CONF_WEBHOOK, []):
```
The default empty list can be moved to the config schema.

braam replied:
Ok, I can change this, will be added in new commit to PR.

MartinHjelmare commented on homeassistant/components/circuit/manifest.json:
```diff
@@ -0,0 +1,8 @@
+  "domain": "circuit",
+  "name": "Unify Circuit",
+  "documentation": "https://www.home-assistant.io/integrations/circuit",
+  "dependencies": [],
```
Empty items can be removed.
```suggestion
```

MartinHjelmare commented on homeassistant/components/circuit/notify.py:
```diff
@@ -0,0 +1,45 @@
+def get_service(hass, config, discovery_info=None):
+    """Get the Unify Circuit notification service."""
+    if discovery_info is None:
+        return
```
```suggestion
        return None
```

braam committed:
Update homeassistant/components/circuit/manifest.json

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

braam committed:
Update homeassistant/components/circuit/notify.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

braam committed:
Update __init__.py

changed empty default list and cv.ensure_list not needed on DOMAIN.
_WEBHOOK_SCHEMA also changed to simply WEBHOKO_SCHEMA

MartinHjelmare commented on homeassistant/components/circuit/__init__.py:
```diff
@@ -0,0 +1,50 @@
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: [
```
Please remove the list here.

braam replied:
list removed, this gives a configuration.yaml change too:
```
circuit:
  webhook:
    - name: circuit_webhook_test
      url: https://eu.yourcircuit.com/rest/v2/webhooks/incoming/<ID-1>
    - name: circuit_hass_dev
      url: https://eu.yourcircuit.com/rest/v2/webhooks/incoming/<ID-2>
    - name: circuit_Webhook_group
      url:
        - https://eu.yourcircuit.com/rest/v2/webhooks/incoming/<ID-1>
        - https://eu.yourcircuit.com/rest/v2/webhooks/incoming/<ID-2>
```

braam committed:
Update __init__.py

removed default list from CONFIG_SCHEMA

MartinHjelmare commented on homeassistant/components/circuit/__init__.py:
```diff
@@ -0,0 +1,45 @@
+    {
+        DOMAIN: vol.Schema(
+            {
+                vol.Optional(CONF_WEBHOOK, default=[]): vol.All(
```
This should probably be required. Otherwise no platform will be set up.

braam replied:
Okay, make sense will change it.

MartinHjelmare commented on homeassistant/components/circuit/__init__.py:
```diff
@@ -0,0 +1,45 @@
+
+async def async_setup(hass, config):
+    """Set up the Unify Circuit component."""
+    domain_config = config.get(DOMAIN)
```
Please use `dict[key]` when we know the key is in the dict.

braam replied:
I don't understand what you mean, but I do check for now if 'url' key is in the dict entry or not before doing the load_platform.

MartinHjelmare replied:
This integration can not be set up without its `DOMAIN` in the config. So we should use `config[DOMAIN]` when accessing the domain config section.

MartinHjelmare commented on homeassistant/components/circuit/notify.py:
```diff
@@ -0,0 +1,45 @@
+    try:
+        return CircuitNotificationService(hass, discovery_info)
+
+    except RuntimeError as err:
```
Why would there be an exception here? It doesn't look possible.

braam replied:
Okay, make sense will change it.

braam committed:
Update notify.py

Removed try --> exception condition

braam committed:
Update __init__.py

CONFIG_SCHEMA changed CONF_WEBHOOK to required.

braam committed:
Update __init__.py

check if 'url' is in the dict entry before proceeding load_platform

MartinHjelmare commented on homeassistant/components/circuit/__init__.py:
```diff
@@ -0,0 +1,46 @@
+
+async def async_setup(hass, config):
+    """Set up the Unify Circuit component."""
+    webhooks = config.get(DOMAIN)[CONF_WEBHOOK]
```
```suggestion
    webhooks = config[DOMAIN][CONF_WEBHOOK]
```

braam replied:
Oh, ok thanks for this now its clear to me.

MartinHjelmare commented on homeassistant/components/circuit/__init__.py:
```diff
@@ -0,0 +1,46 @@
+    {
+        DOMAIN: vol.Schema(
+            {
+                vol.Required(CONF_WEBHOOK, default=[]): vol.All(
```
```suggestion
                vol.Required(CONF_WEBHOOK): vol.All(
```

MartinHjelmare commented on homeassistant/components/circuit/__init__.py:
```diff
@@ -0,0 +1,46 @@
+    webhooks = config.get(DOMAIN)[CONF_WEBHOOK]
+
+    for webhook_conf in webhooks:
+        if 'url' in webhook_conf:
+            hass.async_create_task(
+                discovery.async_load_platform(hass, "notify", DOMAIN, webhook_conf, config)
+            )
```
`CONF_URL` is required in the config schema.
```suggestion
        hass.async_create_task(
            discovery.async_load_platform(hass, "notify", DOMAIN, webhook_conf, config)
        )
```

braam committed:
Update homeassistant/components/circuit/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

braam committed:
Update homeassistant/components/circuit/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

braam committed:
Update homeassistant/components/circuit/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

braam committed:
Update __init__.py

Black formatting has run now.

braam committed:
Update __init__.py

flake8 suggestion: __init__.py:25:83: E231 missing whitespace after ',' change.
black is removeing the space for some unknown reason.

braam commented:
Looks like flake8 and black are fighting each other.. how can I fix this?
flake8: 
`__init__.py:25:83: E231 missing whitespace after ','`

braam committed:
Update __init__.py

Trying without latest',' on line 25, for black and flake8 to have peace.

MartinHjelmare commented on homeassistant/components/circuit/notify.py:
```diff
@@ -0,0 +1,40 @@
+class CircuitNotificationService(BaseNotificationService):
+    """Implement the notification service for Unify Circuit."""
+
+    hass = attr.ib()
```
We don't need `hass`.

braam replied:
Ok, I rewrote the code. Will commit now.

braam committed:
Update notify.py

No hass, no attr used.

braam committed:
Update __init__.py

Changed CONF_URL to string (no list), like suggested a while back. It seems like to be a better idea when the same message has to be posted to multiple channels I can call notify multiple times.

MartinHjelmare commented on homeassistant/components/circuit/__init__.py:
```diff
@@ -0,0 +1,34 @@
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: vol.Schema({vol.Required(CONF_WEBHOOK): [WEBHOOK_SCHEMA]})},
```
Please add `cv.ensure_list` to make sure the webhook schema is wrapped in a list.

braam replied:
Yes, done.

MartinHjelmare commented on homeassistant/components/circuit/notify.py:
```diff
@@ -0,0 +1,38 @@
+class CircuitNotificationService(BaseNotificationService):
+    """Implement the notification service for Unify Circuit."""
+
+    def __init__(self, webhook_url):
```
`config` is a better name since this is a dict with more than one item.

braam replied:
Ok, will change this.

MartinHjelmare commented on homeassistant/components/circuit/__init__.py:
```diff
@@ -0,0 +1,34 @@
+CONF_WEBHOOK = "webhook"
+
+WEBHOOK_SCHEMA = vol.Schema(
+    {vol.Optional(CONF_NAME): cv.string, vol.Required(CONF_URL): cv.string}
```
Where do we use the name item?

braam replied:
It's an optional config entry, when it's used we create the notify service with this name. So it's clear in the config which url will be linked to which conversation.

MartinHjelmare commented on homeassistant/components/circuit/notify.py:
```diff
@@ -0,0 +1,38 @@
+        """Send a message to the webhook."""
+
+        webhook_url = self.webhook_url
+        target = kwargs.get(ATTR_TARGET, webhook_url)
```
The `ATTR_TARGET` item is a list of targets. We're making `target` an inconsistent type now here.

braam replied:
Ok, will remove target.

braam committed:
Update __init__.py

cv.ensure_list added for CONF_WEBHOOK

braam committed:
Update notify.py

Removal target.
Rename to config for dict.

braam committed:
Update homeassistant/components/circuit/notify.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented:
Please link to a docs PR in this PR description. Then we can merge.

braam commented:
Documentation PR: https://github.com/home-assistant/home-assistant.io/pull/13584

MartinHjelmare commented on .coveragerc:
```diff
@@ -122,6 +122,7 @@ omit =
     homeassistant/components/cast/*
     homeassistant/components/cert_expiry/helper.py
     homeassistant/components/channels/*
+    homeassistant/components/circuit/notify.py
```
```suggestion
    homeassistant/components/circuit/*
```

MartinHjelmare committed:
Exclude all circuit modules from .coveragerc
